<a href="https://colab.research.google.com/github/miguelrendon107/Analitica/blob/main/Reto_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**RETO 2**

- Descripcion de las variables:

Si la persona tiene un título universitario(College), si la persona tiene deudas con tarjetas de crédito(CreditCard), el gasto familiar anual en alimentos (FoodSpend), los ingresos anuales (Income), y el gasto anual de los hogares en productos de viaje (TravelSpend).

# ***CARGAR LIBRERIAS Y DATOS***

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt #Graficador
from sklearn.tree import DecisionTreeClassifier

#Para no tener que llamar el archivo
from google.colab import drive
drive.mount('/content/drive')

bd= '/content/drive/MyDrive/ANALITICA NEGOCIOS/2. BD1_Travel_Data_int.xlsx'
BD= pd.read_excel(bd, sheet_name=0)
BD=BD.dropna()
bdx = np.array(BD[['College','FoodSpend', 'TravelSpend', 'Income']])
bdy = np.array(BD['CreditCard']) #Lo que quiero que el modelo pronostique.

print("La base de datos tiene:",len(BD),'registros')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
La base de datos tiene: 500 registros


La base de datos tiene: 500 registros

# ***CREACION DEL MODELO***






In [ ]:
mar= DecisionTreeClassifier(criterion='gini', max_depth=4) #4 niveles o variables de entrada
mar.fit(bdx,bdy)
ydp= mar.predict(bdx)

from sklearn.metrics import confusion_matrix
cm= confusion_matrix(bdy, ydp)
cm
#   0  1
# 0 VN FP
# 1 FN VP
#EVALUAMOS EL MODELO
VN= cm[0,0]; FP= cm[0,1]; FN= cm[1,0]; VP= cm[1,1]

print("Exactitud:",  (VP+VN)/(VP+VN+FP+FN)) #Comportamiento general
print("Sensibilidad:", VP/(FN+VP)) #Como se comporto frente a los positivos
print("Especificidad:", VN/(VN+FP)) #Como se comporto frente a los negativos

from sklearn.tree import export_graphviz
from pydotplus import graph_from_dot_data

vs=['College','FoodSpend', 'TravelSpend', 'Income']
dot_graph= export_graphviz (mar, feature_names=vs)
graph= graph_from_dot_data(dot_graph)
graph.write_png("arbol.png")

Exactitud: 0.898
Sensibilidad: 0.732484076433121
Especificidad: 0.9737609329446064


True

# ***NUEVOS CLIENTES***


In [ ]:
#Calculando las probabilidade de los dos clientes segun el arbol. OPCIÓN 1
BD2= pd.read_excel(bd, sheet_name=1)
BD2=BD2.dropna()
bdx2 = np.array(BD2[['College', 'FoodSpend', 'TravelSpend','Income']])
prob = mar.predict_proba(bdx2)
print(prob)

[[1.         0.        ]
 [0.78676471 0.21323529]]


In [ ]:
#Calculando las probabilidades de los dos clientes segun el arbol. #OPCION 2
#Primer Cliente
#College 0; FoodSpend 2,892.9; Income 65,982; TravelSpend 2151.6

#Información del nodo
Samples = 108
NV= 108
V= 0

print("La probabilidad de que el cliente 1 no use la Tarjeta de Credito es de:",(NV/Samples)*100, "%")
print("La probabilidad de que el cliente 1 use la Tarjeta de Credito es de:",(V/Samples)*100,"%")

#Información del nodo
#Segundo cliente
#College 1; FoodSpend 6,017.6; Income 53,986; TravelSpend 2002.5

Samples = 136
NV= 107
V= 29
print("La probabilidad de que el cliente 2 no use la Tarjeta de Credito es de:",(NV/Samples)*100, "%")
print("La probabilidad de que el cliente 2 use la Tarjeta de Credito es de:",(V/Samples)*100, "%")


La probabilidad de que el cliente 1 no use la Tarjeta de Credito es de: 100.0 %
La probabilidad de que el cliente 1 use la Tarjeta de Credito es de: 0.0 %
La probabilidad de que el cliente 2 no use la Tarjeta de Credito es de: 78.67647058823529 %
La probabilidad de que el cliente 2 use la Tarjeta de Credito es de: 21.323529411764707 %


**Primer cliente:**

La probabilidad de que el cliente 1 no use la Tarjeta de Credito es de: 100.0 %
La probabilidad de que el cliente 1 use la Tarjeta de Credito es de: 0.0 %

**Segundo cliente:**

La probabilidad de que el cliente 2 no use la Tarjeta de Credito es de: 78.67647058823529 %
La probabilidad de que el cliente 2 use la Tarjeta de Credito es de: 21.323529411764707 %

# ***Reglas de Clientes y Nodos Puros***

**Para el cliente 1:**

TravelSpend > 1567.875 AND Income > 59098.0.


**Para el cliente 2:**

Income <= 59098.0 AND FoodSpend > 2711.875 AND TravelSpend <= 3008.25.

**Nodos Puros**

**1**. Nodo [1,0] = TravelSpend <=1188.0 AND FoodSpend > 4972.55.

**2**. Nodo [0,47] = TravelSpend <=1188.0 AND FoodSpend > 4990.73.

**3**. Nodo [108,0] = TravelSpend > 1567.875 AND Income > 59098.0


# **ANALISIS DE RESULTADOS**

El modelo de árbol de decisión, aplicado a 500 registros, alcanzó una exactitud del 89.8%, con una sensibilidad del 73.25% y una especificidad del 97.38%. Esto indica que el modelo es eficaz en predecir correctamente el uso o no de tarjetas de crédito, con un rendimiento particularmente alto en la identificación de aquellos que no las utilizan.

El modelo generó reglas claras que permiten interpretar las decisiones tomadas por el árbol de decisión. Para el primer cliente , con un gasto en viajes superior a 1,567.88 y un ingreso anual superior a 59,098.0, tiene un 100% de probabilidad de no utilizar la tarjeta de crédito. Esto sugiere que, a pesar de tener ingresos relativamente altos y gastar en viajes, este cliente prefiere no recurrir a crédito.

Por otro lardo, para el segundo cliente, , con un ingreso anual inferior a 59,098.0, un gasto en alimentos superior a 2,711.88 y un gasto en viajes inferior a 3,008.25, tiene una probabilidad del 21.32% de utilizar la tarjeta de crédito. Esto sugiere que, aunque con ingresos más bajos, este cliente aún podría recurrir al uso de la tarjeta de crédito, aunque con menor probabilidad que el promedio.

Vale destacar que los nodos puros identificaron que clientes con gastos bajos en viajes y altos en alimentos tienden a no usar tarjetas de crédito.

En conclusión, El modelo es confiable y proporciona reglas claras para segmentar clientes según su uso de tarjetas de crédito.